### PR TITLE
Plane swarm

### PR DESCRIPTION
**Pull request by shreya-kulsh**
This set of code implements follow-mode for ArduPlane. Note that is does not include a robust velocity controller, but should be able to maintain offsets between multiple vehicles.

To test, please use the script plane_formation.sh. This script simulates three vehicles, two of which are on follow mode, at offsets of 5 meters. It is recommended to use MAVProxy with Mission Planner for testing. 
-On MAVProxy, please create a multicast setting by using the command --master mcast: --out=udp:<I.P>:14550
-For further clarification of procedure or current problems with the code, please refer to the document given here: https://drive.google.com/file/d/13ZOd25g6VnXbQV5pKhAYunYRXanLoYAG/view?usp=sharing

### CODE REVIEW

File: ArduPlane/mode_follow.cpp [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+void ModeFollow::update()
+{
+//    gcs().send_text(MAV_SEVERITY_INFO, "follow update %u", millis());
+    Location loc;
+    Vector3f vel_ned;
+    static uint32_t last_notify;
+    if (plane.g2.follow.get_target_location_and_velocity(loc, vel_ned)) {
+        plane.guided_WP_loc = loc;
+        const uint32_t now = AP_HAL::millis();
+        if (now - last_notify > 1000) {
```
Comment (khancyr) [inline ArduPlane/mode_follow.cpp:25]: this is just dead code now !

File: Tools/autotest/arduplane.py [hunk 2]
```diff
@@ -1742,6 +1743,49 @@ def LOITER(self):
...
+                gpi.pack(self.mav.mav)
+                self.mav.mav.send(gpi)
+            self.mav.recv_match(type='GLOBAL_POSITION_INT', blocking=True)
+            pos = self.mav.location()
+            delta = self.get_distance(expected_loc, pos)
+            max_delta = 100
+            self.progress("position delta=%f (want <%f)" % (delta, max_delta))
+            if delta < max_delta:
+                break
         self.fly_home_land_and_disarm()
```
Comment (khancyr) [inline Tools/autotest/arduplane.py:1789]: this shouldn't be removed from the loiter test !

File: Tools/autotest/arduplane.py [hunk 3]
```diff
@@ -1742,6 +1743,49 @@ def LOITER(self):
         self.progress("Centering elevator and ensuring we get back to loiter altitude")
         self.set_rc(2, 1500)
         self.wait_altitude(initial_alt-1, initial_alt+1)
+        
+    def fly_follow_mode(self):
```
Comment (khancyr) [inline Tools/autotest/arduplane.py:1747]: this is the same as copter one. Maybe we could transfer it to common.py

File: libraries/SITL/examples/test/swarm_test1.sh [hunk 4]
```diff
@@ -0,0 +1,35 @@
+#Author:Shreya Kulsh
+#This program tests a formation between two planes
+#!/bin/bash
```
Comment (khancyr) [inline libraries/SITL/examples/test/swarm_test1.sh:3]: same

File: libraries/SITL/examples/test/plane_sim.sh [hunk 5]
```diff
@@ -0,0 +1,37 @@
+#Author:Shreya Kulsh
+#Original producer: Andrew Tridgell
+#!/bin/bash
+
+# three plane swarm (no set formaiton)
```
Comment (khancyr) [inline libraries/SITL/examples/test/plane_sim.sh:5]: this file is useless we could remove it

File: libraries/SITL/examples/test/plane_formation.sh [hunk 6]
```diff
@@ -0,0 +1,47 @@
+#Author:Shreya Kulsh
+#This program tests a formation between three planes
+#!/bin/bash
+
+# three plane swarm (x offset formation)
```
Comment (khancyr) [inline libraries/SITL/examples/test/plane_formation.sh:5]: move the plane into the follow directory, and rename to plane_3_formation or something like that

File: libraries/SITL/examples/test/swarm_test1.sh [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+# two plane swarm (x offset formation)
```
Comment (khancyr) [inline libraries/SITL/examples/test/swarm_test1.sh:5]: move to follow directory and rename to plane_2_follow or something like that

**Review by khancyr — CHANGES_REQUESTED** [review]
Hello, 
Thanks for this, that is nice !
It will need some rework and little corrections ! 
The Follow mode implementation is correct as based on the guided mode. We can live without the velocity controller for now !
The mode implementation is covered by an autotest that is the same the copter one.

The commits need some squashing and cleanup.
We are regrouping commit by feature and files changes.
So you should have as commits : 
- Plane: add Follow mode
- Autotest: move global_position_int_for_location to common.py
- Autotest: plane: add Follow test 
- SITL: add plane follow example scripts
You can use the instructions on https://ardupilot.org/dev/docs/git-interactive-rebase.html for how to do it. Ask on Discord for assistance on git rebasing if you need !

**Comment by EosBandi** [comment]
Apparently, your code does not handle offsets at all. The get_target_location_and_velocity just returns the estimated location and velocity of the lead plane, but then offsets are not added anywhere. So all follow planes just chasing the lead plane, without any offsets.....

**Comment by khancyr** [comment]
Hello,

There is a lot to be done to address review. I will close this for now. Feel free to reopen when you something more ready to merge.
thanks anyway

**Pull request closed without merging**